atoMEC/staticKS.py

```diff
@@ -707,28 +707,84 @@ def calc_E_kin_orbs(eigfuncs, occnums, xgrid):
         E_kin : float
             the kinetic energy
         """
-        # compute the grad^2 component
-        grad2_orbs = mathtools.laplace(eigfuncs, xgrid)
-
-        # compute the (l+1/2)^2 component
-        l_arr = np.array([(l + 0.5) ** 2.0 for l in range(config.lmax)])
-        lhalf_orbs = np.einsum("j,ijkl->ijkl", l_arr, eigfuncs)
-
-        # add together and multiply by eigfuncs*exp(-3x)
-        prefac = np.exp(-3.0 * xgrid) * eigfuncs
-        kin_orbs = prefac * (grad2_orbs - lhalf_orbs)
-
-        # multiply and sum over occupation numbers
-        e_kin_dens = np.einsum("ijk,ijkl->l", occnums, kin_orbs)
+        # compute the kinetic energy density (using default method A)
+        e_kin_dens = Energy.calc_E_kin_dens(eigfuncs, occnums, xgrid)
 
         # FIXME: this is necessary because the Laplacian is not accurate at the boundary
-        e_kin_dens[-3:] = e_kin_dens[-4]
+        for i in range(config.spindims):
+            e_kin_dens[i, -3:] = e_kin_dens[i, -4]
 
         # integrate over sphere
-        E_kin = -0.5 * mathtools.int_sphere(e_kin_dens, xgrid)
+        E_kin = mathtools.int_sphere(np.sum(e_kin_dens, axis=0), xgrid)
 
         return E_kin
 
+    @staticmethod
+    def calc_E_kin_dens(eigfuncs, occnums, xgrid, method="A"):
+        """
+        Calculate the local kinetic energy density (KED).
+
+        There are multiple definitions in the literature of the local KED, see notes.
+
+        Parameters
+        ----------
+        eigfuncs : ndarray
+            the radial KS orbitals on the log grid
+        occnums : ndarray
+            the orbital occupations
+        xgrid : ndarray
+            the logarithmic grid
+        method : str, optional
+            the definition used for KED, can be 'A' or 'B' (see notes).
+
+        Returns
+        -------
+        e_kin_dens : ndarray
+            the local kinetic energy density
+
+        Notes
+        -----
+        The methods 'A' and 'B' in this function are given according to the definitions
+        in the following papers:
+
+        * Hao Jiang, New J. Phys. 22 103050 (2020)
+        * Leon Cohen, J. Chem. Phys. 70, 788 (1979)
+
+        They of course (should) both integrate to the same kinetic energy. The
+        definition 'B' is the one used in the usual definition of the electron
+        localization function (Savin et al., Angewandte Chemie 1992).
+        """
+        if method == "A":
+            # compute the grad^2 component
+            grad2_orbs = mathtools.laplace(eigfuncs, xgrid)
+
+            # compute the (l+1/2)^2 component
+            l_arr = np.array([(l + 0.5) ** 2.0 for l in range(config.lmax)])
```

Comment:
How large is `config.lmax` ? If the number of array elements is high it would be better to avoid creating the list and use a generator comprehension rather than a list comprehension 
```suggestion
            l_arr = np.fromiter(((l + 0.5) ** 2.0 for l in range(config.lmax)), float, config.lmax)
```

Reply:
Not that big (<50), but I see no downside to doing it a better way in any case

Reply:
Ok, in this case... better does not always mean increase in performance, also readability and other things count.;)

Reply:
For me, the new way is no less readable :) In general I'd prefer to use numpy solutions wherever possible in the code so that if/when other people contribute in the future they will hopefully do the same (when it might matter more)